### ISSUE TITLE
Version info/LDFLAGS is missing in postsubmit image push job

### ISSUE THREAD

**Issue by Jont828** [kubernetes-sigs/cluster-api#6098]
**What steps did you take and what happened:**
In the [/post-cluster-api-push-images](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/post-cluster-api-push-images/1491830283345858560) job, the LDFLAGS are missing in the [build logs](https://storage.googleapis.com/kubernetes-jenkins/logs/post-cluster-api-push-images/1491840468340707328/artifacts/build.log). 

The LDFLAGS are computed from `./hack/version.sh`, and it fails because the workspace directory does not have a `.git` folder and we get this error as a result.
```
Step #0: make ARCH=amd64 docker-build-core docker-build-kubeadm-bootstrap docker-build-kubeadm-control-plane 
Step #0: make[3]: Entering directory '/workspace'
Step #0: fatal: not a git repository (or any parent up to mount point /)
Step #0: Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
```

Then, when we are running `docker build`, we can see that the `ldflags` are set to an empty string here

```
Step #0: DOCKER_BUILDKIT=1 docker build --build-arg builder_image=docker.io/library/golang:1.17.3 --build-arg goproxy=https://proxy.golang.org,direct --build-arg ARCH=amd64 --build-arg ldflags="" . -t gcr.io/k8s-staging-cluster-api/cluster-api-controller-amd64:v20220210-v1.1.1
```

**Anything else you would like to add:**
This same issue appears to show up in CAPI and the other Cluster API providers too. I'm working on a [PR in CAPZ](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2068) to try to fix this, and I'd be happy to work together since the same fix could be applied to all the CAP* repos.

**Environment:**

- Cluster-api version: Any
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by sbueringer** [comment]
@Jont828 I was thinking a bit more about this and got curious what `gcr.io/k8s-staging-test-infra/image-builder` is and how it works.

tl;dr is:
> This builder is sugar on top of `gcloud builds submit`. It offers the following features:
> ...
> - Uploading the working directory to GCS once and then reusing it for several builds
> ...
> https://github.com/kubernetes/test-infra/tree/29364b9d3b47613dbd8db503fe2997fcedb1a30c/images/builder#image-builder

While looking at the code I found a `with-git-dir` flag, which disables the exclusion of the .git dir: https://github.com/kubernetes/test-infra/blob/29364b9d3b47613dbd8db503fe2997fcedb1a30c/images/builder/main.go#L140-L142

I found a few jobs which are using that flag in this way: https://github.com/kubernetes/test-infra/blob/669d3a3c0674f87da3f661aff6f301359aee3380/config/jobs/image-pushing/k8s-staging-cluster-api.yaml#L187-L196

What do you think about trying to just enable that flag in `post-cluster-api-provider-azure-push-images` to see if it works?

It looks pretty much like what we need with very minimal effort :)

**Comment by vincepri** [comment]
/milestone v1.2

Let's time the fix before the v1.1.2 release so we can ship it with the proper build version flags.

**Comment by sbueringer** [comment]
@Jont828 
Feel free to assign yourself the issue if you want. I'll just keep track of it / help from the CAPI core side.

**Comment by Jont828** [comment]
@sbueringer Thanks for pointing that out! It looks like the folks at CAPG just merged a [fix with the git arg](https://github.com/kubernetes/test-infra/pull/25182/files). I'll add the change to CAPZ as well to see if it works on our end.

**Comment by sbueringer** [comment]
Looks like this change solved it in CAPG: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/post-cluster-api-provider-gcp-push-images/1493129253376495616

WDYT about opening
* one PR in test-infra to fix CAPI, CAPA, CAPO at the same time (I think we can relatively quick approvals from those providers)
    * cc @mdbooth 
* one PR to solve it for the remaining providers: Nested, IBM Cloud

**Comment by mdbooth** [comment]
I think I'm hearing you don't want me to submit my own CAPO PR for this?

**Comment by sbueringer** [comment]
> I think I'm hearing you don't want me to submit my own CAPO PR for this?

You could, if you want. I would just prefer bundling this a bit, to avoid a bit of noise for folks who are usually very busy / have to approve a lot of PRs :)

**Comment by Jont828** [comment]
@sbueringer I opened a similar [PR](https://github.com/kubernetes/test-infra/pull/25233) with the same fix and it works in CAPZ. I could go ahead and open one to fix it in the rest of the providers as well.

**Comment by Jont828** [comment]
Opened this [PR](https://github.com/kubernetes/test-infra/pull/25251) to fix this in test-infra

**Comment by sbueringer** [comment]
Fixed. Just in time for v1.1.2. Thx @Jont828!

>  Step #0: DOCKER_BUILDKIT=1 docker build --build-arg builder_image=docker.io/library/golang:1.17.3 --build-arg goproxy=https://proxy.golang.org,direct --build-arg ARCH=amd64 --build-arg ldflags="-X 'sigs.k8s.io/cluster-api/version.buildDate=2022-02-17T19:17:24Z' -X 'sigs.k8s.io/cluster-api/version.gitCommit=3433f7b769b4e7f5cb899b2742a5a8a1a9f51b3e' -X 'sigs.k8s.io/cluster-api/version.gitTreeState=clean' -X 'sigs.k8s.io/cluster-api/version.gitMajor=1' -X 'sigs.k8s.io/cluster-api/version.gitMinor=1' -X 'sigs.k8s.io/cluster-api/version.gitVersion=v1.1.2' -X 'sigs.k8s.io/cluster-api/version.gitReleaseCommit=3433f7b769b4e7f5cb899b2742a5a8a1a9f51b3e'" . -t gcr.io/k8s-staging-cluster-api/cluster-api-controller-amd64:v20220217-v1.1.2
> https://storage.googleapis.com/kubernetes-jenkins/logs/post-cluster-api-push-images/1494389837921259520/artifacts/build.log

/close
